init-store failed due to datapath and walpath error for rocksdb backend

Initing HugeGraph Store...
2018-08-22 16:54:58 718   [main] [INFO ] com.baidu.hugegraph.cmd.InitStore [] - Init graph with config file: conf/hugegraph.properties
2018-08-22 16:54:58 819   [main] [INFO ] com.baidu.hugegraph.HugeGraph [] - Opening backend store 'rocksdb' for graph 'hugegraph'
2018-08-22 16:54:58 865   [main] [INFO ] com.baidu.hugegraph.backend.store.rocksdb.RocksDBStore [] - Opening RocksDB with data path: ./rocksdb-data/schema
2018-08-22 16:54:59 1043  [main] [INFO ] com.baidu.hugegraph.backend.store.rocksdb.RocksDBStore [] - Failed to open RocksDB './rocksdb-data/schema' with database 'hugegraph', try to init CF later
2018-08-22 16:54:59 1160  [main] [INFO ] com.baidu.hugegraph.backend.store.rocksdb.RocksDBStore [] - Opening RocksDB with data path: ./rocksdb-data/system
2018-08-22 16:54:59 1217  [main] [INFO ] com.baidu.hugegraph.backend.store.rocksdb.RocksDBStore [] - Failed to open RocksDB './rocksdb-data/system' with database 'hugegraph', try to init CF later
2018-08-22 16:54:59 1276  [main] [INFO ] com.baidu.hugegraph.backend.store.rocksdb.RocksDBStore [] - Opening RocksDB with data path: ./rocksdb-data/graph
2018-08-22 16:54:59 1325  [main] [INFO ] com.baidu.hugegraph.backend.store.rocksdb.RocksDBStore [] - Failed to open RocksDB './rocksdb-data/graph' with database 'hugegraph', try to init CF later
2018-08-22 16:54:59 1409  [main] [INFO ] com.baidu.hugegraph.backend.store.rocksdb.RocksDBStore [] - Opening RocksDB with data path: ./edge_in/graph
2018-08-22 16:54:59 1411  [main] [ERROR] com.baidu.hugegraph.backend.store.rocksdb.RocksDBStore [] - Failed to open RocksDB './edge_in/graph'
        at com.baidu.hugegraph.backend.store.rocksdb.RocksDBStdSessions.<init>(RocksDBStdSessions.java:113) ~[hugegraph-rocksdb-0.7.4.jar:?]
        at com.baidu.hugegraph.backend.store.rocksdb.RocksDBStore.openSessionPool(RocksDBStore.java:206) ~[hugegraph-rocksdb-0.7.4.jar:?]
        at com.baidu.hugegraph.backend.store.rocksdb.RocksDBStore.open(RocksDBStore.java:154) [hugegraph-rocksdb-0.7.4.jar:?]
        at com.baidu.hugegraph.backend.store.rocksdb.RocksDBStore.open(RocksDBStore.java:143) [hugegraph-rocksdb-0.7.4.jar:?]
        at com.baidu.hugegraph.HugeGraph.initBackend(HugeGraph.java:187) [hugegraph-core-0.7.4.jar:0.7.4.0]
        at com.baidu.hugegraph.cmd.InitStore.initBackend(InitStore.java:115) [hugegraph-dist-0.7.4.jar:?]
        at com.baidu.hugegraph.cmd.InitStore.initGraph(InitStore.java:103) [hugegraph-dist-0.7.4.jar:?]
        at com.baidu.hugegraph.cmd.InitStore.main(InitStore.java:86) [hugegraph-dist-0.7.4.jar:?]
Exception in thread "main" com.baidu.hugegraph.backend.BackendException: Failed to open RocksDB './edge_in/graph'
        at com.baidu.hugegraph.backend.store.rocksdb.RocksDBStore.open(RocksDBStore.java:186)
        at com.baidu.hugegraph.backend.store.rocksdb.RocksDBStore.open(RocksDBStore.java:143)
        at com.baidu.hugegraph.HugeGraph.initBackend(HugeGraph.java:187)
        at com.baidu.hugegraph.cmd.InitStore.initBackend(InitStore.java:115)
        at com.baidu.hugegraph.cmd.InitStore.initGraph(InitStore.java:103)
        at com.baidu.hugegraph.cmd.InitStore.main(InitStore.java:86)
Caused by: org.rocksdb.RocksDBException: While lock file: ./rocksdb-data/graph/LOCK: No locks available
        at org.rocksdb.RocksDB.open(Native Method)
        at org.rocksdb.RocksDB.open(RocksDB.java:286)
        at com.baidu.hugegraph.backend.store.rocksdb.RocksDBStdSessions.<init>(RocksDBStdSessions.java:113)
        at com.baidu.hugegraph.backend.store.rocksdb.RocksDBStore.openSessionPool(RocksDBStore.java:206)
        at com.baidu.hugegraph.backend.store.rocksdb.RocksDBStore.open(RocksDBStore.java:154)
        ... 5 more

--- a/hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBSessions.java
+++ b/hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBSessions.java
@@ -19,14 +19,10 @@
 
 package com.baidu.hugegraph.backend.store.rocksdb;
 
-import java.io.IOException;
-import java.nio.file.Paths;
 import java.util.Set;
 
-import org.apache.commons.io.FileUtils;
 import org.rocksdb.RocksDBException;
 
-import com.baidu.hugegraph.backend.BackendException;
 import com.baidu.hugegraph.backend.store.BackendEntry.BackendColumnIterator;
 import com.baidu.hugegraph.backend.store.BackendSession;
 import com.baidu.hugegraph.backend.store.BackendSessionPool;
@@ -47,21 +43,6 @@ public abstract class RocksDBSessions extends BackendSessionPool {
 
     @Override
     public abstract Session session();
-
-    public String wrapPath(String path) {
-        return wrapPath(path, this.store);
-    }
-
-    public static String wrapPath(String path, String store) {
-        // Ensure the `path` exists
-        try {
-            FileUtils.forceMkdir(FileUtils.getFile(path));
-        } catch (IOException e) {
-            throw new BackendException(e.getMessage(), e);
-        }
-        // Join with store type
-        return Paths.get(path, store).toString();
-    }
 
     /**
      * Session for RocksDB

--- a/hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStdSessions.java
+++ b/hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStdSessions.java
@@ -62,14 +62,12 @@ public class RocksDBStdSessions extends RocksDBSessions {
     private final HugeConfig conf;
     private final RocksDB rocksdb;
 
-    public RocksDBStdSessions(HugeConfig config, String database, String store)
+    public RocksDBStdSessions(HugeConfig config, String dataPath,
+                              String walPath, String database, String store)
                               throws RocksDBException {
         super(database, store);
 
         this.conf = config;
-
-        String dataPath = wrapPath(this.conf.get(RocksDBOptions.DATA_PATH));
-        String walPath = wrapPath(this.conf.get(RocksDBOptions.WAL_PATH));
 
         // Init options
         Options options = new Options();
@@ -83,13 +81,11 @@ public class RocksDBStdSessions extends RocksDBSessions {
         this.rocksdb = RocksDB.open(options, dataPath);
     }
 
-    public RocksDBStdSessions(HugeConfig config, String database, String store,
+    public RocksDBStdSessions(HugeConfig config, String dataPath,
+                              String walPath, String database, String store,
                               List<String> cfNames) throws RocksDBException {
         super(database, store);
         this.conf = config;
-
-        String dataPath = wrapPath(this.conf.get(RocksDBOptions.DATA_PATH));
-        String walPath = wrapPath(this.conf.get(RocksDBOptions.WAL_PATH));
 
         // Old CFs should always be opened
         List<String> cfs = this.mergeOldCFs(dataPath, cfNames);

--- a/hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java
+++ b/hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java
@@ -19,6 +19,8 @@
 
 package com.baidu.hugegraph.backend.store.rocksdb;
 
+import java.io.IOException;
+import java.nio.file.Paths;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.HashMap;
@@ -29,6 +31,7 @@ import java.util.Map.Entry;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Collectors;
 
+import org.apache.commons.io.FileUtils;
 import org.rocksdb.RocksDBException;
 import org.slf4j.Logger;
 
@@ -129,9 +132,10 @@ public abstract class RocksDBStore implements BackendStore {
         }
 
         // Open base disk
-        String dataPath = config.get(RocksDBOptions.DATA_PATH);
-        dataPath = RocksDBSessions.wrapPath(dataPath, this.store);
-        this.sessions = this.open(config, dataPath, this.tableNames());
+        String dataPath = this.wrapPath(config.get(RocksDBOptions.DATA_PATH));
+        String walPath = this.wrapPath(config.get(RocksDBOptions.WAL_PATH));
+
+        this.sessions = this.open(config, dataPath, walPath, this.tableNames());
 
         // Open tables with optimized disk
         List<String> disks = config.get(RocksDBOptions.DATA_DISKS);
@@ -140,18 +144,19 @@ public abstract class RocksDBStore implements BackendStore {
             for (Entry<HugeType, String> e : this.tableDiskMapping.entrySet()) {
                 String table = this.table(e.getKey()).table();
                 String disk = e.getValue();
-                this.open(config, disk, Arrays.asList(table));
+                this.open(config, disk, disk, Arrays.asList(table));
             }
         }
     }
 
     protected RocksDBSessions open(HugeConfig config, String dataPath,
-                                   List<String> tableNames) {
+                                   String walPath, List<String> tableNames) {
         LOG.info("Opening RocksDB with data path: {}", dataPath);
 
         RocksDBSessions sessions = null;
         try {
-            sessions = this.openSessionPool(config, tableNames);
+            sessions = this.openSessionPool(config, dataPath,
+                                            walPath, tableNames);
         } catch (RocksDBException e) {
             if (dbs.containsKey(dataPath)) {
                 if (e.getMessage().contains("No locks available")) {
@@ -162,7 +167,8 @@ public abstract class RocksDBStore implements BackendStore {
                     try {
                         // Will open old CFs(of other keyspace)
                         final List<String> none = ImmutableList.of();
-                        sessions = this.openSessionPool(config, none);
+                        sessions = this.openSessionPool(config, dataPath,
+                                                        walPath, none);
                     } catch (RocksDBException e1) {
                         // Let it throw later
                         e = e1;
@@ -174,7 +180,8 @@ public abstract class RocksDBStore implements BackendStore {
                          "try to init CF later", dataPath, this.database);
                 try {
                     // Only open default CF, won't open old CFs
-                    sessions = this.openSessionPool(config, null);
+                    sessions = this.openSessionPool(config, dataPath,
+                                                    walPath, null);
                 } catch (RocksDBException e1) {
                     LOG.error("Failed to open RocksDB with default CF", e1);
                 }
@@ -198,12 +205,15 @@ public abstract class RocksDBStore implements BackendStore {
     }
 
     protected RocksDBSessions openSessionPool(HugeConfig config,
+                                              String dataPath, String walPath,
                                               List<String> tableNames)
                                               throws RocksDBException {
         if (tableNames == null) {
-            return new RocksDBStdSessions(config, this.database, this.store);
+            return new RocksDBStdSessions(config, dataPath, walPath,
+                                          this.database, this.store);
         } else {
-            return new RocksDBStdSessions(config, this.database, this.store,
+            return new RocksDBStdSessions(config, dataPath, walPath,
+                                          this.database, this.store,
                                           tableNames);
         }
     }
@@ -401,7 +411,7 @@ public abstract class RocksDBStore implements BackendStore {
             String store = pair[0].trim();
             HugeType table = HugeType.valueOf(pair[1].trim().toUpperCase());
             if (this.store.equals(store)) {
-                path = RocksDBSessions.wrapPath(path, this.store);
+                path = this.wrapPath(path);
                 this.tableDiskMapping.put(table, path);
             }
         }
@@ -412,6 +422,17 @@ public abstract class RocksDBStore implements BackendStore {
         E.checkState(db != null && !db.closed(),
                      "RocksDB store has not been opened: %s", disk);
         return db;
+    }
+
+    protected String wrapPath(String path) {
+        // Ensure the `path` exists
+        try {
+            FileUtils.forceMkdir(FileUtils.getFile(path));
+        } catch (IOException e) {
+            throw new BackendException(e.getMessage(), e);
+        }
+        // Join with store type
+        return Paths.get(path, this.store).toString();
     }
 
     /***************************** Store defines *****************************/

--- a/hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdbsst/RocksDBSstSessions.java
+++ b/hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdbsst/RocksDBSstSessions.java
@@ -22,12 +22,8 @@ package com.baidu.hugegraph.backend.store.rocksdbsst;
 import java.io.File;
 import java.nio.file.Path;
 import java.nio.file.Paths;
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
 import java.util.Map.Entry;
-import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 
 import org.apache.commons.lang3.tuple.Pair;
@@ -38,7 +34,6 @@ import org.rocksdb.SstFileWriter;
 
 import com.baidu.hugegraph.backend.BackendException;
 import com.baidu.hugegraph.backend.store.BackendEntry.BackendColumnIterator;
-import com.baidu.hugegraph.backend.store.rocksdb.RocksDBOptions;
 import com.baidu.hugegraph.backend.store.rocksdb.RocksDBSessions;
 import com.baidu.hugegraph.backend.store.rocksdb.RocksDBStdSessions;
 import com.baidu.hugegraph.config.HugeConfig;
@@ -51,11 +46,12 @@ public class RocksDBSstSessions extends RocksDBSessions {
     private final String dataPath;
     private final Map<String, SstFileWriter> tables;
 
-    public RocksDBSstSessions(HugeConfig conf, String database, String store) {
+    public RocksDBSstSessions(HugeConfig conf, String dataPath,
+                              String database, String store) {
         super(database, store);
 
         this.conf = conf;
-        this.dataPath = this.wrapPath(this.conf.get(RocksDBOptions.DATA_PATH));
+        this.dataPath = dataPath;
         this.tables = new ConcurrentHashMap<>();
 
         File path = new File(this.dataPath);
@@ -64,9 +60,10 @@ public class RocksDBSstSessions extends RocksDBSessions {
         }
     }
 
-    public RocksDBSstSessions(HugeConfig config, String database, String store,
+    public RocksDBSstSessions(HugeConfig config, String dataPath,
+                              String database, String store,
                               List<String> tableNames) throws RocksDBException {
-        this(config, database, store);
+        this(config, dataPath, database, store);
         for (String table : tableNames) {
             this.createTable(table);
         }

--- a/hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdbsst/RocksDBSstStore.java
+++ b/hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdbsst/RocksDBSstStore.java
@@ -40,13 +40,14 @@ public abstract class RocksDBSstStore extends RocksDBStore {
 
     @Override
     protected RocksDBSessions openSessionPool(HugeConfig config,
+                                              String dataPath, String walPath,
                                               List<String> tableNames)
                                               throws RocksDBException {
         if (tableNames == null) {
-            return new RocksDBSstSessions(config, this.database(),
+            return new RocksDBSstSessions(config, dataPath, this.database(),
                                           this.store());
         } else {
-            return new RocksDBSstSessions(config, this.database(),
+            return new RocksDBSstSessions(config, dataPath, this.database(),
                                           this.store(), tableNames);
         }
     }

--- a/hugegraph-test/src/main/java/com/baidu/hugegraph/unit/rocksdb/BaseRocksDBUnitTest.java
+++ b/hugegraph-test/src/main/java/com/baidu/hugegraph/unit/rocksdb/BaseRocksDBUnitTest.java
@@ -34,7 +34,6 @@ import org.junit.Before;
 import org.mockito.Mockito;
 import org.rocksdb.RocksDBException;
 
-import com.baidu.hugegraph.backend.store.rocksdb.RocksDBOptions;
 import com.baidu.hugegraph.backend.store.rocksdb.RocksDBSessions;
 import com.baidu.hugegraph.backend.store.rocksdb.RocksDBStdSessions;
 import com.baidu.hugegraph.config.HugeConfig;
@@ -117,9 +116,8 @@ public class BaseRocksDBUnitTest extends BaseUnitTest {
         Configuration conf = Mockito.mock(PropertiesConfiguration.class);
         Mockito.when(conf.getKeys()).thenReturn(Collections.emptyIterator());
         HugeConfig config = new HugeConfig(conf);
-        config.setProperty(RocksDBOptions.DATA_PATH.name(), DB_PATH);
-        config.setProperty(RocksDBOptions.WAL_PATH.name(), DB_PATH);
-        RocksDBSessions rocks = new RocksDBStdSessions(config, "db", "store");
+        RocksDBSessions rocks = new RocksDBStdSessions(config, DB_PATH, DB_PATH,
+                                                       "db", "store");
         rocks.createTable(table);
         return rocks;
     }

